packages/arb-token-bridge-ui/src/hooks/useCCTP.ts

```diff
@@ -0,0 +1,254 @@
+}
+
+export function useCCTP({
+  sourceChainId,
+  walletAddress
+}: {
+  sourceChainId?: CCTPSupportedChainId
+  walletAddress?: `0x${string}` | string
+}) {
+  const {
+    tokenMessengerContractAddress,
+    targetChainDomain,
+    targetChainId,
+    attestationApiUrl,
+    usdcContractAddress,
+    messengerTransmitterContractAddress
+  } = getContracts(sourceChainId)
+  const { data: usdcToken } = useToken({ address: usdcContractAddress })
+
+  const destinationAddress = useMemo(() => {
+    if (!walletAddress) return
+
+    // CCTP uses 32 bytes addresses, while EVEM uses 20 bytes addresses
+    return utils.hexlify(utils.zeroPad(walletAddress, 32)) as `0x${string}`
+  }, [walletAddress])
+
+  const depositForBurn = useCallback(
+    async (amount: BigNumber) => {
+      const config = await prepareWriteContract({
+        address: tokenMessengerContractAddress,
+        abi: tokenMessengerAbi,
+        functionName: 'depositForBurn',
+        args: [
+          amount,
+          targetChainDomain,
+          destinationAddress as `0x${string}`,
+          usdcContractAddress
+        ]
+      })
+      return writeContract(config)
+    },
+    [
+      tokenMessengerContractAddress,
+      targetChainDomain,
+      destinationAddress,
+      usdcContractAddress
+    ]
+  )
+
+  const fetchAttestation = useCallback(
+    async (attestationHash: `0x${string}`) => {
+      const response = await fetch(
+        `${attestationApiUrl}/attestations/${attestationHash}`
+      )
+
+      const attestationResponse: AttestationResponse = await response.json()
+      return attestationResponse
+    },
+    [attestationApiUrl]
+  )
+
+  const waitForAttestation = useCallback(
+    async (attestationHash: `0x${string}`) => {
+      while (true) {
+        const attestation = await fetchAttestation(attestationHash)
+        if (attestation.status === 'complete') {
+          return attestation.attestation
+        }
+
+        await new Promise(r => setTimeout(r, 5000))
+      }
+    },
+    [fetchAttestation]
+  )
+
+  const receiveMessage = useCallback(
+    async ({
+      messageBytes,
+      attestation
+    }: {
+      messageBytes: `0x${string}`
+      attestation: `0x${string}`
+    }) => {
+      const config = await prepareWriteContract({
+        address: messengerTransmitterContractAddress,
+        abi: messengerTransmitterAbi,
+        functionName: 'receiveMessage',
+        chainId: targetChainId,
+        args: [messageBytes, attestation]
+      })
+      return writeContract(config)
+    },
+    [messengerTransmitterContractAddress, targetChainId]
+  )
+
+  const approveUSDC = useCallback(
+    async (signer: Signer, amount: BigNumber) => {
+      const contract = ERC20__factory.connect(usdcContractAddress, signer)
+      return contract.functions.approve(tokenMessengerContractAddress, amount)
+    },
+    [usdcContractAddress, tokenMessengerContractAddress]
+  )
+
+  const approveAndDeposit = useCallback(
```

Comment:
We chatted about this on Slack, this can be a separate hook that's one level of abstraction higher (it uses `approveForBurn` and `depositForBurn` from `useCCTP` internally). We can do it in another PR.